# Título del Proyecto: Ciencia de Datos aplicada en la Industria Retail

## Breve descripción del proyecto (descripción del dataset, problemas interesantes asociados)
Se provee de un dataset que contiene datos históricos de ventas correspondientes a 45 tiendas anónimas localizadas en diferentes departamentos. Además el dataset cuenta con información de contexto (temperatura, precio del combustible, tasa de desempleo, información de rebajas, feriados) para las fechas correspondientes a las ventas (ventas registradas semanalmente desde febrero de 2010 hasta noviembre de 2012). Se provee también de una caracterización por tipo y tamaño para cada tienda y una lista de los usuarios que frecuentan cada una.
Uno de los desafíos de modelar datos de esta industria se presenta debido a la necesidad de tomar decisiones basadas en operaciones históricas limitadas. Se busca la definición de estrategias y toma de decisiones, en base al análisis y procesamiento de los datos históricos disponibles, para el cumplimiento de un cierto objetivo, como lo puede ser aumento de la rentabilidad del negocio/mejora del servicio prestado al cliente.
El objetivo de este proyecto es que el estudiante a lo largo de las materias a cursar en la diplomatura identifique y aplique diferentes técnicas de análisis/procesamiento de los datos que generen información valiosa para un negocio que se desarrolla en la industria en cuestión. Algunos de los puntos interesantes asociados son:
- Predicción de ventas futuras (detección de altas y bajas), análisis de estacionalidad de los datos.
- Segmentación de las ventas en base a las características de su contexto.
- Sistema de recomendación de tiendas para los usuarios.
- Análisis del efecto de las rebajas en las ventas para las distintas tiendas.
- Análisis del impacto de los feriados en las ventas.
- Análisis de correlación entre las diferentes variables provistas y las ventas semanales, análisis del impacto que las mismas causan.

## Análisis y Visualización
- Correlación entre variables/análisis de independencia. A través de este análisis se puede ver el impacto de cada variable sobre las ventas semanales y de esta manera determinar cuales son las variables importantes a considerar.
- Distribución de los ejemplos con respecto a las diferentes clases. 
- Análisis de outliers.
- Visualización de las ventas con respecto al tiempo para cada tienda en un determinado departamento.
- Visualización de las ventas totales de cada tienda.
- Porcentaje que representan las rebajas sobre las ventas totales para una tienda.

## Análisis y Curación
Para análisis y curación podrán aplicarse sobre el dataset (en su totalidad) los siguientes puntos:
- Importación de datos.
- Chequeo de claves únicas por sample/eliminar duplicados.
- Despersonalización de datos.
- Normalización de los nombres de las columnas en los dataframes.
- Tratamiento de valores faltantes.
- Codificación de variables categóricas.
- Análisis de valores atípicos.
- Persistencia de los resultados.
- Ordenamiento de las columnas.
- Eliminar columnas que no aporten información.
- Crear un dataset único de las 3 tablas provistas incluyendo toda la información útil en una misma tabla.

## Introducción al ML
Los dataset provistos contienen muchas variables sobre las cuales se puede aplicar análisis y procesamiento. En esta materia puntual lo importante es que el estudiante aprenda a hacer la division de los datos, la elección del modelo, evaluación de metricas, independientemente de la complejidad del dataset. Razón por la cual para este práctico se tomaran dos variables simples de analizar, por ejemplo la fecha de las ventas y la columna que nos dice si es feriado o no, y con esto hacer un sistema predictivo en donde se introduce una fecha y el modelo predice si en esa semana hay feriados o no. Con esto podrá aplicarse:
- Carga de datos.
- Una pequeña reestructuración de las columnas optimizandolo para el análisis que se desea hacer (por ejemplo considerar todas las fechas de la semana y no solo la provista por el dataset).
- Division en conjuntos de entrenamiento y evaluación.
- Elección de un modelo.
- Selección de hyperparámetros.
- Métricas sobre el conjunto de evaluación.
- Curvas ROC.

## Aprendizaje Supervisado
En este caso, en la materia se van aplicando diferentes técnicas correspondientes a aprendizaje supervisado evaluando los resultados obtenidos. Lo ideal es partir de un baseline e ir complejizando el modelo. Sería interesante implementar en este caso un sistema predictivo de ventas para un determinado año en base a la información histórica. Se cuenta con features como la fecha (de la que se puede derivar mes o epoca del año), precio del combustible, desempleo, temperatura las cuales se pueden utilizar como entrada de un modelo que prediga las sales que se tendrá en una determinada fecha. Esto podría hacerse como un modelo general, o seleccionar una determinada tienda para hacer el análisis. Se cuenta con el registro de 3 años, la idea sería entrenar el modelo con los dos primeros y testear el comportamiento del modelo con los datos del tercero. Es decir, se entrena con 2010 y 2011, y se corre el modelo con las features del 2012 comparando la salida del predictor con la real.

## Aprendizaje No Supervisado
Para la aplicación de aprendizaje no supervisado sería interesante hacer dos análisis:
- Segmentación de las ventas: clusterizar las ventas junto con las diferentes variables disponibles y hacer un análisis de los grupos obtenidos. Si aplica, evaluar en que caso tenemos las mejores ventas, ante que condiciones, con que combinación de variables. Podría utilizarse un algoritmo como K Means con algún método de optimización de número de clusters como elbow method.
- Sistema de recomendación de tiendas: se cuenta con el dato de los usuarios que frecuentan cada tienda, puede armarse un pequeño sistema de recomendación de tiendas a usuarios a través de métodos matriciales.

## Optativas Relacionadas
Análisis de series temporales, Introducción al data Warehousing, Ciencia de datos en las Finanzas, Introducción al aprendizaje profundo.

## Contenido del Dataset
Se provee de datos históricos de ventas de 45 tiendas localizadas en diferentes regiones, cada tienda contiene un número determinado de departamentos. Cada compañía además organiza eventos de rebajas a lo largo del año. Estas rebajas preceden a feriados destacados como lo son: el Super Bowl, el Día del Trabajo, Día de Acción de Gracias y Navidad. Las semanas que incluyen estos feriados tienen un peso 5 veces mayor que las que no.

El contenido de los cuatro archivos (Stores, Users, Features y Sales) se presenta a continuación:

### Stores
Información anónima de las 45 tiendas, incluye tipo y tamaño de las mismas.

### Users
Lista de usuarios y las tiendas que frecuentan.

### Features
Contiene información adicional relacionada a las tiendas, la localidad y actividad regional para las fechas dadas.
- Store: el número de tienda.
- Date: la semana correspondiente.
- Temperatura: temperatura promedio de la región.
- Fuel Price: costo del combustible en la región.
- Markdown 1-5: datos anónimos relacionados a rebajas promocionales. Esta información esta sólo disponible a partir de Noviembre del 2011 y no esta disponible para todas las tiendas todo el tiempo. Lo valores faltantes están marcados como NA.
- CPI: índice de precios al consumidor.
- Unemployment: tasa de desempleo.
- IsHoliday: si la semana contiene al menos un día feriado.

### Sales
Datos historicos de ventas, cubren desde el 05-02-2010 hasta el 01-11-2012. Incluye los siguientes campos:
- Store: el número de tienda.
- Dept: el número de departamento.
- Date: la fecha correspondiente a la transacción.
- Weekly_Sales: ventas para el departamento dado en la tienda correspondiente.
- IsHoliday: indica si corresponde a una semana con días feriados o no.

Link a la fuente: https://www.kaggle.com/manjeetsingh/retaildataset#Features%20data%20set.csv

## Lectura y breve introducción al manejo del Dataset
Nótese que por simplicidad se mostrarán siempre 10 ejemplos de cada tabla usando comandos como head o sample.

Primero importamos la librería pandas, a través de la cual realizaremos la carga de los datos en DataFrames

In [1]:
import pandas as pd

Cargamos los archivos contenedores del dataset en cuatro DataFrames distintos

In [2]:
features = pd.read_csv('features.csv')
sales = pd.read_csv('sales.csv')
stores = pd.read_csv('stores.csv')
users = pd.read_csv('users.csv')

Visualizamos los diez primeros valores de cada DataFrame

In [3]:
features.head(10)

Unnamed: 0  Store        Date  Temperature  Fuel_Price  MarkDown1  \
0           0      1  05/02/2010        42.31       2.572        NaN   
1           1      1  12/02/2010        38.51       2.548        NaN   
2           2      1  19/02/2010        39.93       2.514        NaN   
3           3      1  26/02/2010        46.63       2.561        NaN   
4           4      1  05/03/2010        46.50       2.625        NaN   
5           5      1  12/03/2010        57.79       2.667        NaN   
6           6      1  19/03/2010        54.58       2.720        NaN   
7           7      1  26/03/2010        51.45       2.732        NaN   
8           8      1  02/04/2010        62.27       2.719        NaN   
9           9      1  09/04/2010        65.86       2.770        NaN   

   MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  \
0        NaN        NaN        NaN        NaN  211.096358         8.106   
1        NaN        NaN        NaN        NaN  211.242170         8.106   
2        NaN        NaN        NaN        NaN  211.289143         8.106   
3        NaN        NaN        NaN        NaN  211.319643         8.106   
4        NaN        NaN        NaN        NaN  211.350143         8.106   
5        NaN        NaN        NaN        NaN  211.380643         8.106   
6        NaN        NaN        NaN        NaN  211.215635         8.106   
7        NaN        NaN        NaN        NaN  211.018042         8.106   
8        NaN        NaN        NaN        NaN  210.820450         7.808   
9        NaN        NaN        NaN        NaN  210.622857         7.808   

   IsHoliday  
0      False  
1       True  
2      False  
3      False  
4      False  
5      False  
6      False  
7      False  
8      False  
9      False

In [4]:
sales.head(10)

Unnamed: 0  Store  Dept        Date  Weekly_Sales IsHoliday  Sale Id
0           0      1     1  05/02/2010      24924.50     False        0
1           1      1     1  12/02/2010      46039.49      True        1
2           2      1     1  19/02/2010      41595.55     False        2
3           3      1     1  26/02/2010      19403.54     False        3
4           4      1     1  05/03/2010      21827.90     False        4
5           5      1     1  12/03/2010      21043.39     False        5
6           6      1     1  19/03/2010      22136.64     False        6
7           7      1     1  26/03/2010      26229.21     False        7
8           8      1     1  02/04/2010      57258.43     False        8
9           9      1     1  09/04/2010      42960.91     False        9

In [5]:
stores.head(10)

Unnamed: 0  Store Type      Size
0           0      1    A  151315.0
1           1      2    A  202307.0
2           2      3    B   37392.0
3           3      4    A       NaN
4           4      5    B   34875.0
5           5      6    A  202505.0
6           6      7    B   70713.0
7           7      8    A  155078.0
8           8      9    B  125833.0
9           9     10    B  126512.0

In [6]:
users.head(10)

Unnamed: 0              users          stores
0           0       Sara Arriaga      16-32-7-44
1           1      Tomas Citelli     20-37-10-11
2           2   Sebastian Araujo  34-14-18-16-29
3           3  Horacio Villareal     22-14-39-25
4           4          Juan Mira   31-7-27-21-13
5           5       Gustavo Mora      45-7-19-18
6           6     Carlos Citelli   43-24-33-5-13
7           7        Franco Mira   44-12-20-8-45
8           8     Fede Villareal        19-45-32
9           9   Noelia Villareal     5-37-6-18-9

En caso de querer ver las columnas que contiene un DataFrame se puede hacer:

In [7]:
sales.columns.values

array(['Unnamed: 0', 'Store', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday',
       'Sale Id'], dtype=object)

En caso de querer cambiar el nombre de una columna:

In [8]:
sales.rename(columns={'Sale Id':'Sale_Id'}, inplace=True)
sales.head(10)

Unnamed: 0  Store  Dept        Date  Weekly_Sales IsHoliday  Sale_Id
0           0      1     1  05/02/2010      24924.50     False        0
1           1      1     1  12/02/2010      46039.49      True        1
2           2      1     1  19/02/2010      41595.55     False        2
3           3      1     1  26/02/2010      19403.54     False        3
4           4      1     1  05/03/2010      21827.90     False        4
5           5      1     1  12/03/2010      21043.39     False        5
6           6      1     1  19/03/2010      22136.64     False        6
7           7      1     1  26/03/2010      26229.21     False        7
8           8      1     1  02/04/2010      57258.43     False        8
9           9      1     1  09/04/2010      42960.91     False        9

Si quisieramos ver el tipo de datos:

In [9]:
sales.dtypes

Unnamed: 0        int64
Store             int64
Dept              int64
Date             object
Weekly_Sales    float64
IsHoliday        object
Sale_Id           int64
dtype: object

Para ver un resumen de nuestro DataFrame simplemente:

In [10]:
sales.describe()

Unnamed: 0          Store           Dept   Weekly_Sales  \
count  425796.000000  425796.000000  425796.000000  425796.000000   
mean   212897.500000      22.200035      44.260944   15980.254676   
std    122916.861947      12.785342      30.494688   22711.970177   
min         0.000000       1.000000       1.000000   -4988.940000   
25%    106448.750000      11.000000      18.000000    2080.495000   
50%    212897.500000      22.000000      37.000000    7610.830000   
75%    319346.250000      33.000000      74.000000   20204.122500   
max    425795.000000      45.000000      99.000000  693099.360000   

             Sale_Id  
count  425796.000000  
mean   210779.929558  
std    121698.163454  
min         0.000000  
25%    105381.750000  
50%    210804.500000  
75%    316161.250000  
max    421569.000000

Ademas se pueden aplicar diversas operaciones sobre columnas, como por ejemplo el cálculo del promedio y la varianza:

In [11]:
print(sales['Weekly_Sales'].mean())
print(sales['Weekly_Sales'].var())

15980.254675877844
515833589.33176196


Una librería muy útil para el cálculo de estadísticos y diferentes operaciones matemáticas es la librería numpy, a continuación se realiza el cálculo del promedio a modo de ejemplo:

In [12]:
import numpy as np

sales_array = np.array(sales['Weekly_Sales'])
sales_array.mean()

15980.254675877648

Por defecto no podremos ver todas las columnas y filas de los DataFrames en caso que sean muy extensos. Para poder hacerlo deberemos setear los siguientes parametros de pandas:

In [13]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

Cómo se puede observar en todas las tablas tenemos una columna Unnamed: 0 que no nos aporta información. Podemos borrar columnas de un DataFrame de la siguiente manera:

In [14]:
sales = sales.drop('Unnamed: 0', axis=1)
sales.head(10)

Store  Dept        Date  Weekly_Sales IsHoliday  Sale_Id
0      1     1  05/02/2010      24924.50     False        0
1      1     1  12/02/2010      46039.49      True        1
2      1     1  19/02/2010      41595.55     False        2
3      1     1  26/02/2010      19403.54     False        3
4      1     1  05/03/2010      21827.90     False        4
5      1     1  12/03/2010      21043.39     False        5
6      1     1  19/03/2010      22136.64     False        6
7      1     1  26/03/2010      26229.21     False        7
8      1     1  02/04/2010      57258.43     False        8
9      1     1  09/04/2010      42960.91     False        9

Para visualizar filas podremos hacerlo de la siguiente manera:

In [15]:
sales['Weekly_Sales'].head(10)

0    24924.50
1    46039.49
2    41595.55
3    19403.54
4    21827.90
5    21043.39
6    22136.64
7    26229.21
8    57258.43
9    42960.91
Name: Weekly_Sales, dtype: float64

O simplemente (en este caso al usar sample en lugar de head estamos mostrando 10 elementos random del dataset):

In [16]:
sales.Date.sample(10)

144504    04/03/2011
57967     02/04/2010
175482    05/10/2012
275427    04/11/2011
387789    17/06/2011
31077     27/07/2012
393737    06/04/2012
37596     14/09/2012
211912    19/10/2012
261113    08/07/2011
Name: Date, dtype: object

Si quisieramos quedarnos con todas las filas correspondientes a los días feriados podríamos hacerlo de la siguiente manera:

In [17]:
sales[sales['IsHoliday']==True].head(10)

Store  Dept        Date  Weekly_Sales IsHoliday  Sale_Id
1        1     1  12/02/2010      46039.49      True        1
31       1     1  10/09/2010      18194.74      True       31
42       1     1  26/11/2010      18820.29      True       42
47       1     1  31/12/2010      19124.58      True       47
83       1     1  09/09/2011      17746.68      True       83
94       1     1  25/11/2011      20911.25      True       94
99       1     1  30/12/2011      23350.88      True       99
105      1     1  10/02/2012      36988.49      True      105
135      1     1  07/09/2012      18322.37      True      135
144      1     2  12/02/2010      44682.74      True      144

Si se quiere recorrer todas las filas de un DataFrame podrá utilizarse el comando iterrows, que devolverá la información para cada fila de la siguiente manera:
(0, Store                    1
Dept                     1
Date            05/02/2010
Weekly_Sales       24924.5
IsHoliday            False
Sale Id                  0
Name: 0, dtype: object)

Y nos permitirá acceder a los datos a través del número de la columna correspondiente, en este caso imprimiremos todas las ventas correspondientes a la Store 1 en el departamento 1 en días feriados:

In [18]:
for row in sales.iterrows():
    if row[1][0] == 1 and row[1][1] == 1 and row[1][4]==True:
        print(row[1][3])

46039.49
18194.74
18820.29
19124.58
17746.68
20911.25
23350.88
36988.49
18322.37
17746.68


Muchas veces en nuestros Datasets tenemos datos vacíos (NaN), los cuales pueden ser molestos para nuestros análisis. En un análisis profundo se debe analizar cada caso y completar los valores vacíos según ciertos criterios. Por ejemplo una fila de edad vacía podría completarse con el promedio del resto de las edades. En caso de un análisis rápido podrían simplemente eliminarse las filas (o columnas) que contienen valores NaN de la siguiente manera:

In [19]:
features.dropna(axis=0, how='any').head(10)

Unnamed: 0  Store        Date  Temperature  Fuel_Price  MarkDown1  \
92           92      1  11/11/2011        59.11       3.297   10382.90   
93           93      1  18/11/2011        62.25       3.308    6074.12   
94           94      1  25/11/2011        60.14       3.236     410.31   
95           95      1  02/12/2011        48.91       3.172    5629.51   
96           96      1  09/12/2011        43.93       3.158    4640.65   
97           97      1  16/12/2011        51.63       3.159    5011.32   
98           98      1  23/12/2011        47.96       3.112    2725.36   
99           99      1  30/12/2011        44.55       3.129    5762.10   
100         100      1  06/01/2012        49.01       3.157    6277.39   
101         101      1  13/01/2012        48.53       3.261    5183.29   

     MarkDown2  MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  \
92     6115.67     215.07    2406.62    6551.42  217.998085         7.866   
93      254.39      51.98     427.39    5988.57  218.220509         7.866   
94       98.00   55805.51       8.00     554.92  218.467621         7.866   
95       68.00    1398.11    2084.64   20475.32  218.714733         7.866   
96       19.00     105.02    3639.42   14461.82  218.961846         7.866   
97       67.00     347.37     225.79    4011.37  219.179453         7.866   
98       40.48     634.70      24.90    2739.43  219.357722         7.866   
99    46011.38     260.36     983.65    4735.78  219.535990         7.866   
100   21813.16     143.10    1450.13    8483.00  219.714258         7.348   
101    8025.87      42.24     453.08    3719.38  219.892526         7.348   

     IsHoliday  
92       False  
93       False  
94        True  
95       False  
96       False  
97       False  
98       False  
99        True  
100      False  
101      False

Cuando queramos aplicar una funcion a la totalidad de elementos de una columna podremos hacerlo de la siguiente manera (en este caso reemplazaremos los valores nan de la columna ventas por el promedio):

In [20]:
def replace_nan(val):
    if str(val)=='NaN':
        print('Nan founded')

Veámos ahora la distribución de los valores de las clases de una columna en particular:

In [21]:
sales.IsHoliday.value_counts()

False    356626
True      27052
Name: IsHoliday, dtype: int64

Y lo comparamos con el tamaño del dataframe:

In [22]:
sales.shape

(425796, 6)

Cómo se puede observar no coincide la suma de ambas clases con el número total de filas, probablemente tengamos valores nulos, los cuales los podemos reemplazar con el valor más frecuente de esta columna:

In [23]:
sales.IsHoliday.value_counts().idxmax()

False

In [24]:
def replace_nan(val):
    if str(val)=='nan':
        return sales.IsHoliday.value_counts().idxmax()
    else:
        return val

In [25]:
sales.IsHoliday.head(10).apply(replace_nan)

0    False
1     True
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: IsHoliday, dtype: bool

## Consigna práctico Análisis y Visualización

- Calcular estadísticos como la moda, media, mediana y desviación estándar del precio del combustible y la temperatura. ¿Responden a alguna distribución conocida?
- Seleccionar una tienda cualquiera, y calcular el promedio de ventas mensuales para los años en cuestión, graficar la distribución de las ventas promedios mensuales para cada año. ¿Responde a alguna distribución conocida?
- Realizar un análisis de outliers para 3 variables a elección.
- ¿Qué pasaría con las ventas si se baja el combustible?¿Que pasaría con las ventas si aumenta la tasa de desempleo?¿Qué sucede con las ventas si nos encontramos en un día feriado? En el caso de las ventas mensuales promedios, ¿existe una relación entre la variable ventas y el mes del año en que nos encontramos?
- Crear una columna adicional IsMarkdown la cual será True si ha habido una rebaja en esa fecha y será False si no la ha habido. Teniendo en cuenta las variables IsHolisday e IsMarkdown, calcular su probabilidades conjunta y marginal.
- ¿Que probabilidad hay de que haya rebajas una semana que se sabe que es feriado? Dividir en 4 partes el dataset y calcular bayes con respecto a estas dos variables, usando los resultados de cada iteración / partición para calcular. El objetivo es simular que los datos que van llegando en cada iteración recalculan la probabilidad

Para la entrega se pide un jupyter notebook que contenga el desarrollo y los cálculos que han sido necesario para responder las preguntas. Además el alumno deberá presentar las respuestas acompañadas de gráficos en un formato interactivo para el lector, como lo puede ser a través de un html.

## Consigna práctico Análisis y Curación: Limpiando un Dataset
1. Importando los datos:
    - Verificar si no hay problemas en la importación: importar los datos, visualizarlos, análisis de tipos, corrección en los tipos de los datos de entrada.
    - Asegurar que el archivo sales posee Ids/Claves únicas. Para el resto de los archivos, ¿tenemos algún atributo que se comporte como clave única?, en caso positivo chequear que no se repite. En caso de no tener una clave única identificatoria, ¿sería relevante asignar una clave única a cada registro?, chequear que no existen datos duplicados para estos casos.
    - Despersonalizar los datos y guardarlos en un nuevo archivo. Tener en cuenta nunca modificar los datos crudos u originales.
2. Pasos necesarios para limpieza del dataset:
    - Etiquetas de variables/columnas: no usar caracteres especiales. Verificar que no haya problemas de codificación/encoding.
    - Tratar valores faltantes (NaN).
    - Codificar variables: las variables categóricas deberán ser tratadas como variables numéricas.
    - Verificar la consistencia de las variables: constatar que los valores de cada atributo tienen sentido, detectar valores que no son consistentes con el resto.
    - Identificar valores atípicos en nuestro dataset. ¿Qué es conveniente hacer con ellos? Evaluar cada caso.
    - Juntar las columnas de interés en un mismo DataFrame (Sales con Features). 
    - Para simplificar el problema resamplear los datos ¿Transformar Weekly sales en ventas mensuales?. Graficar la distribución de las ventas mensuales para cada año para 5 tiendas a elección. Comparar sus distribuciones. ¿Se reconoce alguna distribución conocida?
    - Analizar correlación entre número departamento y ventas semanales/mensuales, ¿posee alguna incidencia el número de departamento sobre las ventas?, en caso negativo eliminar esta variable de nuestros dataframes.
    - Una vez que tenemos las features de interés de nuestro set de datos, aplicar algún método de normalización sobre los mismos, para evitar tener un sesgo de unas variables sobre otra (se pueden utilizar técnicas como z-score/min-max scaling). Guardar el dataset normalizado con un nombre representativo.
    - Finalmente, reducir los features de interés mediante sus combinaciones lineales (aplicando Principal Component Analysis). Guardar el dataset con nombre representativo.
    - Enumere formas eficientes de guardado y acceso de grandes volúmenes de datos.
    - Guardar todos los archivos depurados con nombres representativos.

Bonus: entregar el práctico corriendo en una imágen de Docker.

Material de lectura recomendado:

https://www.machinelearningplus.com/time-series/time-series-analysis-python/

https://towardsdatascience.com/playing-with-time-series-data-in-python-959e2485bff8


https://sebastianraschka.com/Articles/2014_about_feature_scaling.html

## Consigna Práctico Introducción ML:
En esta materia puntual lo importante es que el estudiante aprenda a hacer la division de los datos, la elección del modelo, evaluación de metricas, independientemente de la complejidad del dataset. Es por esto que se toman para el análisis variables simples del dataset. De esta manera el estudiante puede hacer foco en los conceptos que se desean fijar.

Las variables a tratar serán:
### Grupo Alpha:
- Fecha y columna IsHoliday. El objetivo será construir un predictor que tome la fecha (tratada según criterio de los estudiantes) y prediga si en esa semana hay un feriado o no.

### Grupo Omega:
- Ventas, IsHoliday y columna IsMarkdown. El objetivo es tomar como entrada el valor de las ventas para una semana y la información de si tiene un día feriado, y predecir si hubo rebajas esa semana.

Actividades:
1. Importaciones. Seteo de semilla para reproducibilidad. Carga del Dataset. Selección y armado del subconjunto de las variables deseadas.
2. División de los datos en conjunto de entrenamiento y test. El conjunto de entrenamiento se utilizará para la creación y selección de los clasificadores. Una vez elegido el model, este set también los servirá para la selección de los hiperparámetros. Una vez entrenado el modelo, se utilizará el conjunto de test (conjunto de datos independiente) para la evaluación del modelo final.
3. Elección de un modelo. En este punto se deberán seguir los siguientes pasos:
    - Selección de hipótesis: elegir la función/modelo que se utilizará para la implementación del clasificador.
    - Selección del regularizador: analizar si es mejor regularizar con módulo o por cuadrados.
    - Selección de la función de costo con la cual se va a optimizar el modelo.
    - Para todos los casos justificar las selecciones.
4. Selección de hiperparámetros. Luego de seleccionar y setear el modelo que se utilizará, variar sus hiperparámetros y quedarse con la mejor combinación.
5. Evaluación del modelo final. Una vez seleccionados los hiperparámetros y entrenado el modelo, correrlo sobre el conjunto de test/evaluación. Utilizar las métricas que se crean convenientes. Siempre es bueno evaluar con más de una métrica. Que significa la acuracy de un modelo? En que se diferencian Precision y Recall? Cuando es conveniente usar cada una de estas métricas?
6. Es el umbral por default el mejor umbral de decisión? En caso contrario, cual es el valor del umbral óptimo?

## Ejemplo para selección de Modelo:
clfs = [
    KNeighborsClassifier(),
    MultinomialNB(),
    DecisionTreeClassifier(max_depth = 2, random_state=0),
    LogisticRegression(random_state=0),
    LinearSVC(random_state=0),
    SVC(random_state=0),
    RandomForestClassifier(random_state=0),
]

vect = CountVectorizer(binary=True)

for clf in clfs:
    print(str(clf.__class__))
    pipeline = Pipeline([
        ('vect', vect),
        ('clf', clf),
    ])
    pipeline.fit(X_train, y_train)
    
    y_pred = pipeline.predict(X_train)
    print(metrics.classification_report(y_train, y_pred, target_names=['accuweather',
                                                                       'allianz',
                                                                       'amazon',
                                                                       'citigroup',
                                                                       'fujitsu',
                                                                       'garmin',
                                                                       'github',
                                                                       'ibm',
                                                                       'metlife',
                                                                       'uber',
                                                                       'yahoo',
                                                                       'zurich']))

    y_pred = pipeline.predict(X_test)
    print(metrics.classification_report(y_test, y_pred, target_names=['accuweather',
                                                                       'allianz',
                                                                       'amazon',
                                                                       'citigroup',
                                                                       'fujitsu',
                                                                       'garmin',
                                                                       'github',
                                                                       'ibm',
                                                                       'metlife',
                                                                       'uber',
                                                                       'yahoo',
                                                                       'zurich']))

Ejemplo para seleccion de parámetros:

param_grid = {
    'vect__binary': [True],
    'vect__lowercase': [True],
    'vect__sublinear_tf': [True, False],
    'vect__ngram_range': [(1, 3), (1, 4),(1, 7)],
    'vect__strip_accents': ['ascii'],
    'vect__analyzer': ['word'],
    'vect__min_df': [1, 2, 3],
    'vect__max_df': [1.],
    'clf__multi_class' : ['ovr', 'crammer_singer'],
    'clf__random_state': [0],
    'clf__fit_intercept':[True, False],
    'clf__loss':['hinge', 'squared_hinge'],
    'clf__C':[1.0, 0.1],
}

params_list = list(ParameterGrid(param_grid))

pipeline = Pipeline([
    ('vect', TfidfVectorizer()),
    ('clf', LinearSVC()),
])

results = []
print(len(params_list))

bar = progressbar.ProgressBar(maxval=len(params_list), 
                              widgets=[progressbar.Bar('=', '[', ']'), ' ', 
                                       progressbar.Percentage()])
bar.start()


i = 0
barra = ''
for params in params_list:
    bar.update(i+1)
    pipeline.set_params(**params)
    pipeline.fit(X_train, y_train)
    
    y_pred = pipeline.predict(X_test)
    acc = metrics.accuracy_score(y_test, y_pred)
    f1 = metrics.f1_score(y_test, y_pred, average='macro')
    result = {'acc': acc, 'f1': f1}
        
    results.append({
        **result,
        **params,
    })
    i += 1
bar.finish()

results_df = pd.DataFrame(results)
results_df.sort_values(['acc', 'f1'], ascending=False)[:10]

# Consigna Práctico Aprendizaje Supervisado
En este trabajo, se utilizarán herramientas y conocimientos adquiridos en materias previas para la preparación de los datos y selección de parámetros pero se hará foco en la evaluación de diferentes modelos de apredizaje supervisado para la resolución de un problema más complejo.

La idea es que ambos grupos trabajen con variables diferentes para poder luego hacer una comparación entre ambos resultados:
### Grupo Alpha:
- Construir un modelo que prediga el valor de ventas (sales) en base a las features disponibles sin utilizar la fecha (usar numero de store, departamento, temperatura, tasa de desempleo, cpi, precio del combustible, columna is holiday y columna is markdown).

### Grupo Omega:
- Construir un modelo que prediga el valor de las ventas (sales), utilizando las variables anteriores, más el valor de la fecha (pre tratada correspondientemente).

Luego de estos dos análisis podremos analizar como afecta la variable fecha a nuestros resultados. Si bien se considera la variable fecha, no se está teniendo en cuenta la secuencia de los datos como lo haríamos en un análisis de series temporales. Simplemente tomamos el valor día, mes, año sin evaluar la secuencia de los datos que tenemos. Este modelo no es el ideal para este tipo de análisis, pero es el adecuado para aplicar en base a los conocimientos vistos en la materia.

## Actividades:
1. Cargar el dataset, quedarse con las columnas de interés, division de los datos en conjuntos de train, test.
2. Buscar la mejor representación para nuestras variables de entrada.
2. Armar un baseline para resolver inicialmente nuestro problema (un modelo simple, como una regresión lineal, el cual nos sirve para evaluar como van mejorando nuestros resultados).
3. ¿Cuáles son las mejores métricas para evaluar nuestros modelos predictivos?
4. Armar modelos más complejos (logistic regression, SVMs, decision trees), entrenarlos y evaluar el comportamiento de cada uno con la métrica definida en 3. Quedarse con el modelo que mejor ajuste.
5. Evaluar el mejor conjunto de hiperparámetros para el modelo elegido.
6. Correr las métricas de evaluación contra el modelo final. Analizar puntualmente los casos en los que la predicción difiere mucho del valor real, ¿Qué sucede en estos casos?

Bonus track (opcional): correr una LSTM con la columna fechas y sales como salidas. Las LSTM si tienen en cuenta secuencia de las fechas y probablemente mejoren significativamente los resultados.

Link útil para corrida de LSTM: https://towardsdatascience.com/playing-with-time-series-data-in-python-959e2485bff8

# Consigna Práctico Aprendizaje No Supervisado

En este práctico se aplicarán algoritmos de clustering a los datos que venimos trabajando. La mayor parte del esfuerzo no va a estar en que algoritmo de clustering aplicar, sino en el trabajo/selección de los datos de entrada, variación de parámetros y análisis de las soluciones obtenidas. Se trabajará sobre dos ejes centrales. Uno será la agrupación de los datos en base a las features para poder analizar si las mismas nos ayudan a inferir el rango por el cual se mueven las ventas. El otro tratará de la agrupación de los usuarios en base a las tiendas que frecuentan, y la creación de un sistema de recomendación en base a los perfiles encontrados luego de la agrupación.

## 1. Preparación de Dataset:

### Grupo alpha

Tomar los datos del archivo Sales y armar un Dataset con las siguientes características:

- Fecha de la semana correspondiente a la venta.
- Ventas. Para un análisis mas sencillo pasaremos la Weekly Sale a variable categórica. Teniendo en cuenta los valores máximos y mínimos de venta los dividiremos de forma tal que nos queden 3 categorías: venta baja, media y alta.
- Store.
- Departamento.
- Temperatura, precio del combustible, CPI, tasa de desempleo, es feriado o no, hubo rebaja o no correspondientes a cada fecha.

### Grupo omega:

Tomar los datos del archivo Usuarios. Se deberá pasar la variable de las tiendas que frecuentan los usuarios a un vector tipo "One Hot" para poder procesarlo. Es decir se tendrá:

- Nombre del usuario (desperzonalizado).
- Vector con dimension igual al número de tiendas, con un 1 sobre las tiendas que frecuenta el usuario y un 0 sobre las que no frecuenta.
    
## 2. Aplicación, optimización y evaluación de modelo de aprendizaje no supervisado:

### Grupo alpha:

- Tomar las features (sin fecha y sin ventas) y correr un algoritmo de clustering (recomendado K Means), buscando encontrar un patrón por el cual podamos agrupar los datos. 
- Correr el algoritmo para diferentes valores de número de clusters.
- ¿Cómo podríamos encontrar un valor óptimo de número de clusters? Encontrar dicho valor.
- ¿Se ha logrado tener clusters con cantidades similares de elementos? Análisis de los resultados en función del balanceo que presentan los clusters.
- Evaluar los resultados con la medida de pureza tomando a la variable categórica Ventas para la evaluación de la pureza dentro de cada cluster. ¿Hemos logrado encontrar un patrón que defina nuestras ventas?
- Aplicar PCA para disminuir la cantidad de features y evaluar si este cambio arroja mejores resultados.

### Grupo omega:

- Tomar el vector correspondiente a las tiendas y correr un algoritmo de clustering (recomendado K Means), buscando encontrar un patrón por el cual podamos agrupar los datos. 
- Correr el algoritmo para diferentes valores de número de clusters. 
- ¿Cómo podríamos encontrar un valor óptimo de número de clusters? Encontrar dicho valor. 
- ¿Se ha logrado tener clusters con cantidades similares de elementos? Análisis de los resultados en función del balanceo que presentan los clusters.
- Crear un sistema de recomendación en donde dado un usuario nuevo que entra al sistema, se lo asigna a un perfil de usuario y se le recomiendan las tiendas que corresponden a su perfil y que aún no ha visitado.